docs/architecture/adr-001-transfer-module-implementation.md

```diff
@@ -0,0 +1,177 @@
+necessity stems from the operational structure of `ibc-rs` handlers, where the
+processing of ICS-20 packets relies on core handlers accessible exclusively
+through `Ibc`.
+
+Upon receiving a message, `Ibc` proceeds to identify its type, such as ICS-20
+packets, and directs it to the corresponding module. Specifically, in our setup,
+this entails invoking the
+[`transfer`](https://github.com/informalsystems/sovereign-ibc/blob/0c3b99f44613ff9a8668ade798b39507a17a7321/modules/sov-ibc/src/call.rs#L66)
+method to route the ICS-20 messages to the `IbcTransfer` for subsequent handling
+and processing.
+
+## Module Structure
+
+In addition to its core functionalities, `IbcTransfer` maintains a record of
+tokens minted on a rollup created by the transfer module. You can see detailed
+of this structure
+[here](https://github.com/informalsystems/sovereign-ibc/blob/4e37dc4bb88624765384d1662549c00e991acc4a/modules/sov-ibc-transfer/src/lib.rs#L20-L50)
+in the `sov-ibc-transfer` crate.
+
+Specifically, `IbcTransfer` manages two essential maps:
+
+* `minted_token_name_to_id`: This map links the token name to its corresponding
+  token ID for tokens created by IBC. It is used during minting and burning
+  processes to check if that the token exists and to obtain the necessary ID for
+  these operations.
+
+* `minted_token_id_to_name`: This map connects the token ID to its corresponding
+  token name for tokens created by IBC. It is utilized during escrow and
+  un-escrow processes to confirm that the `TokenId` obtained from the `denom` is
+  **not** an IBC-created token, indicating it is a native token for escrow and
+  un-escrow operations.
+
+If the purpose of these maps is not entirely clear, reviewing how each transfer
+scenario is handled in the following section will provide clarity.
+
+## Transfer Scenarios
+
+Given the context provided, the `IbcTransfer` module handles four scenarios,
+each comprising validation *(x_validate method)* and execution *(x_execute
+method)* stages.
+
+Before diving into the specifics of each scenario, it is essential to understand
+that the token name on the Sovereign SDK rollups is not guaranteed to be unique,
+and hence when transferring **native tokens**  we must use the token ID (which
+is guaranteed to be unique) as the ICS-20 denom to ensure uniqueness.
+
+### Escrowing Tokens - Sender on Rollup with Rollup as Source
+
+1. Verify that the `memo` field does not exceed maximum allowed length to
+   prevent large memos overwhelming the system. We set the maximum memo length
+   to 32768 bytes like the `ibc-go`.
+2. Identify the token ID by parsing the `denom` field of the receiving
+   `MsgTransfer`.
+3. Validate that the token is native and **not** an IBC-created token by
+   cross-referencing with the `minted_token_id_to_name` state.
+   * If the token ID is found in the `minted_token_id_to_name` state, it is an
+     IBC-created token, and the transfer is rejected.
```

Comment:
I am now thinking if this check should be relaxed. Because we are escrowing the ibc denoms that didn't come from the channel end corresponding to the current ics20 transaction.

Reply:
Very good point! We can check if the channel end corresponds to the passed pair of port-channel IDs. 
I've addressed this in PR: https://github.com/informalsystems/sovereign-ibc/pull/138